storage/redistore/peer_store.go

```diff
@@ -0,0 +1,169 @@
+//Note: ip6 seperation into shards is unnecessary when using Redis(?)
+package redistore
+
+import (
+	"encoding/binary"
+	"log"
+	"time"
+
+	"github.com/chihaya/chihaya/bittorrent"
+	"github.com/chihaya/chihaya/storage"
+	"github.com/garyburd/redigo/redis"
+)
+
+var namespacePrefix string
+
+type Config struct {
+	Namespace                 bool          `yaml:"namespace"`
+	Cntrl                     string        `yaml:"prefix"`
+	MaxNumWant                int           `yaml:"max_numwant"`
+	Host                      string        `yaml:"host"`
+	Port                      string        `yaml:"port"`
+	GarbageCollectionInterval time.Duration `yaml:"gc_interval"`
+	PeerLifetime              time.Duration `yaml:"peer_liftetime"`
+}
+
+type peerStore struct {
+	conn         redis.Conn
+	closed       chan struct{}
+	maxNumWant   int
+	peerLifetime time.Duration
+	gcValidity   int
+}
+
+func New(cfg Config) (storage.PeerStore, error) {
+	conn, err := redis.Dial("tcp", cfg.Host+":"+cfg.Port)
+	if err != nil {
+		log.Fatal("Connection failed:" + err.Error())
+		return nil, err
+	}
+
+	if !cfg.Namespace {
+		conn.Do("SELECT", cfg.Cntrl)
+	} else {
+		namespacePrefix = cfg.Cntrl
+	}
+
+	ps := &peerStore{
+		conn:         conn,
+		closed:       make(chan struct{}),
+		maxNumWant:   cfg.MaxNumWant,
+		peerLifetime: cfg.PeerLifetime,
+		gcValidity:   int(cfg.GarbageCollectionInterval.Seconds()),
+	}
+
+	return ps, nil
+}
+
+type serializedPeer string
+
+func newPeerKey(p bittorrent.Peer) serializedPeer {
+	b := make([]byte, 20+2+len(p.IP))
+	copy(b[:20], p.ID[:])
+	binary.BigEndian.PutUint16(b[20:22], p.Port)
+	copy(b[22:], p.IP)
+
+	return serializedPeer(b)
+}
+
+func panicIfClosed(closed <-chan struct{}) {
+	select {
+	case <-closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+}
+
+func addNameSpace(command string) string {
+	if namespacePrefix != "" {
+		return namespacePrefix + ":" + command
+	}
+	return command
+}
+
+func (s *peerStore) PutSeeder(infoHash bittorrent.InfoHash, p bittorrent.Peer) error {
+	panicIfClosed(s.closed)
+
+	pk := newPeerKey(p)
+	return addAndCleanPeer(s, infoHash, "seeder", pk)
+}
+
+func (s *peerStore) DeleteSeeder(infoHash bittorrent.InfoHash, p bittorrent.Peer) error {
+	panicIfClosed(s.closed)
+	pk := newPeerKey(p)
+	return remPeers(s, infoHash, "seeder", pk)
+}
+
+func (s *peerStore) PutLeecher(infoHash bittorrent.InfoHash, p bittorrent.Peer) error {
+	panicIfClosed(s.closed)
+	pk := newPeerKey(p)
+	return addAndCleanPeer(s, infoHash, "leecher", pk)
+
+}
+
+func (s *peerStore) DeleteLeecher(infoHash bittorrent.InfoHash, p bittorrent.Peer) error {
+	panicIfClosed(s.closed)
+	pk := newPeerKey(p)
+	return remPeers(s, infoHash, "leecher", pk)
+
+}
+
+func (s *peerStore) GraduateLeecher(infoHash bittorrent.InfoHash, p bittorrent.Peer) error {
+	panicIfClosed(s.closed)
+	err := s.DeleteLeecher(infoHash, p)
+	if err != nil {
+		return err
+	}
+	err = s.PutSeeder(infoHash, p)
```

Comment:
probably you should do this first and delete leecher considering the possibility of the application getting crashed between these two steps.